base/strings/lazy.jl

```diff
@@ -0,0 +1,35 @@
+"""
+    LazyString <: AbstractString
+
+A lazy representation of string interpolation. This is useful when a string
+needs to be constructed in a context where performing the actual interpolation
+and string construction is unnecessary or undesirable (e.g. in error paths
+of functions).
+
+This type is designed to be cheap to construct at runtime, trying to offload
+as much work as possible to either the macro or later printing operations.
+"""
+struct LazyString <: AbstractString
```

Comment:
It feels a bit cheeky to declare that this is a kind of String, then implement none of the interface for one.

We could potentially make this object mutable, then swap the field for the string'd value when you first access it, perhaps?

Reply:
Yes, I think that's reasonable. Let me do that.